### ISSUE TITLE
Kconfig: Build-time warning and minimum version requirement for Clang

### ISSUE THREAD

**Issue by dileks** [ClangBuiltLinux/linux#941]
Just saw these commits in `peterz/queue.git#locking/atomics` [1]:

"compiler/gcc: Emit build-time warning for GCC prior to version 4.8"
"compiler/gcc: Raise minimum GCC version for kernel builds to 4.8"

...and what do people think to do this analog for Clang (`CC_IS_CLANG`) to what there is for `CC_IS_GCC` and `GCC_VERSION` and move/improve the checks to/in `scripts/Kconfig.include` and add the minimum requirement of `CLANG_VERSION`?

We had a discussion/brainstorming about that on the ClangBuiltLinux meetup in Zurich.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/peterz/queue.git/log/?h=locking/atomics
[2] https://lkml.kernel.org/r/20200123153341.19947-2-will@kernel.org
[3] https://lkml.kernel.org/r/20200123153341.19947-10-will@kernel.org

**Comment by dileks** [comment]
Analog in `include/linux/compiler-clang.h`:
```
[ include/linux/compiler-gcc.h ]

/*
 * Common definitions for all gcc versions go here.
 */
#define GCC_VERSION (__GNUC__ * 10000		\
		     + __GNUC_MINOR__ * 100	\
		     + __GNUC_PATCHLEVEL__)

/* https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58145 */
#if GCC_VERSION < 40800
# error Sorry, your compiler is too old - please upgrade it.
#endif
```

https://git.kernel.org/pub/scm/linux/kernel/git/peterz/queue.git/tree/include/linux/compiler-gcc.h?h=locking/atomics

**Comment by dileks** [comment]
According to `scripts/clang-version.sh`:
```
MAJOR=$(echo __clang_major__ | $compiler -E -x c - | tail -n 1)
MINOR=$(echo __clang_minor__ | $compiler -E -x c - | tail -n 1)
PATCHLEVEL=$(echo __clang_patchlevel__ | $compiler -E -x c - | tail -n 1)
```
...this should result in `include/linux/compiler-clang.h` with minimum requirement version 9.0.0:
```
/*
 * Common definitions for all clang versions go here.
 */
#define CLANG_VERSION (__clang_major__ * 10000 \
		       __clang_minor__ * 100   \
		       __clang_patchlevel__)

#if CLANG_VERSION < 90000
# error Sorry, your compiler is too old - please upgrade it.
#endif
```

**Comment by dileks** [comment]
Sample code snippet on top what is in `peterz/queue.git#locking/atomic` (for demonstration reasons I have included the GCC changes):
```
[ init/Kconfig ]

 config CC_IS_GCC
-	def_bool $(success,$(CC) --version | head -n 1 | grep -q gcc)
+	def_bool $(cc-is-gcc)
 
 config GCC_VERSION
 	int
-	default $(shell,$(srctree)/scripts/gcc-version.sh $(CC)) if CC_IS_GCC
+	default $(gcc-version) if CC_IS_GCC

config CC_IS_CLANG
-   default $(shell,$(srctree)/scripts/clang-version.sh $(CC))
+   def_bool $(cc-is-clang)

config CLANG_VERSION
	int
-   def_bool $(success,$(CC) --version | head -n 1 | grep -q clang)
+   default $(clang-version) if CC_IS_CLANG


[ scripts/Kconfig.include ]

+# gcc version including patch level
+gcc-version := $(shell,$(srctree)/scripts/gcc-version.sh $(CC))

+# Return y if the compiler is GCC, n otherwise
+cc-is-gcc := $(success,$(CC) --version | head -n 1 | grep -q gcc)

+# clang version including patch leve
+clang-version := $(shell,$(srctree)/scripts/clang-version.sh $(CC))

+# Return y if the compiler is Clang, n otherwise
+cc-is-clang := $(success,$(CC) --version | head -n 1 | grep -q clang)

NOTE: Might be better cc-is-xxx before cc-version ananlog to `init/Kconfig` (see above).

[ Documentation/process/changes.rst ]

 ====================== ===============  ========================================
         Program        Minimal version       Command to check the version
 ====================== ===============  ========================================
 
+Clang                  9.0              clang --version
```

**Comment by dileks** [comment]
Another options is to put such informations in `Documentation/kbuild/llvm.rst` (see commit 
`Documentation/llvm: add documentation on building w/ Clang/LLVM`).

[1] https://git.kernel.org/pub/scm/linux/kernel/git/masahiroy/linux-kbuild.git/commit/?h=for-next&id=fcf1b6a35c16ac500fa908a4022238e5d666eabf

**Comment by nathanchance** [comment]
I'll comment here for now rather than on the email thread that you started. I think that the technical implementation of this is completely benign and should not be the focus of discussion. I think we need to discuss the meta/logistics of this level of change. Some of my initial thoughts to spur further discussion...

### Somewhat objective brain dump

By setting a minimum supported version, we have to be committed to keep it building with that version. Look at [all of the code](https://git.kernel.org/pub/scm/linux/kernel/git/peterz/queue.git/commit/?id=06e693e3fc3a809f966a46b11550cc4bfe60611a) that is being ripped out. We would have to be willing to insert similar hacks or workarounds to keep whatever minimum version we choose working properly or we have to be okay with constantly bumping the minimum version whenever we fix a severe enough bug.

Historically, the kernel community's attitude towards GCC bugs is report them and pray they eventually get fixed while inserting a hack working around it (or raising the minimum version like in this case). It seems like the kernel community's attitude towards clang bugs has been "fix your compiler"; in this scheme, it becomes more complicated. The core kernel community would have to take clang seriously for this to work properly.

### Completely opinionated brain dump

I was initially pro-minimum clang version but as I have thought more about it, I do not believe we are ready for this change. We are still working on adding architectures. We are still working on the integrated assembler. We are still working on getting warnings down to zero, which may still require further clang changes. We do not have a ton of contributors to help with issues, which would undoubtedly grow if we are stamping our seal that we are "ready" to be in `Documentation/process/changes.rst`.

Furthermore, if we are being completely realistic, we do not really have great testing in place for making sure that all of this stuff stays working. Our CI is limited by Travis's build capacity. KernelCI is using `clang-9`, which is great, but they are not entirely within our control and they are not testing everything for us. 0day is using tip of tree (which is great so we can avoid breakage up there) but they give us no coverage of older clang versions. Something that I want to do with my next two weeks off (yay, coronavirus...) is create a testing framework to qualify a release version of clang, which would build a version of clang from source, build every known working config, and boot it if need be, which we could use every time that Hans/Tom release a new LLVM release candidate. By extension, we could use that framework to find a suitable minimum version.

If we want to give some sort of indication of what clang version _should_ work for the kernel, I would say we should add it to `Documentation/kbuild/llvm.rst` (which might actually be a better place for this kind of information anyways?) and keep it soft: "LLVM 9 and higher is known to work better due to the implementation of such features such as `asm goto`".

**Comment by dileks** [comment]
Thanks @nathanchance for your thoughts.

As said `Documentation/kbuild/llvm.rs` is currently the best place to document the current situation.

Anyway, I want to get some feedback here on other architectures than x86-64 what I am using.

Also, some feedback on the LLVM utilities like LLD & Co.

Clang/LLVM version 10 will be released hopefully next week which will be more "suitable" to build current versions of the Linux-kernel.

**Comment by dileks** [comment]
Something like this patch (on top of Linux v5.6-rc7).
[kconfig-Rework-compiler-identification-and-version-handling.txt](https://github.com/ClangBuiltLinux/linux/files/4394330/kconfig-Rework-compiler-identification-and-version-handling.txt)

**Comment by dileks** [comment]
Re-tested both patches with clang-10 and Linux v5.6 by lowering version check to v9.0.0 as minimum.

**Comment by dileks** [comment]
Added patches on top of kbuild-next (RFC status!).

```
[PATCH] kconfig: Rework compiler identification and version handling
[PATCH] compiler/clang: Emit build-time warning for Clang prior to version 9.0.0
```

Note: I use this patchset with my experiments starting with Linux v5.6-rc7.

[0001-kconfig-Rework-compiler-identification-and-version-h_patch.txt](https://github.com/ClangBuiltLinux/linux/files/4460246/0001-kconfig-Rework-compiler-identification-and-version-h_patch.txt)
[0002-compiler-clang-Emit-build-time-warning-for-Clang-pri_patch.txt](https://github.com/ClangBuiltLinux/linux/files/4460247/0002-compiler-clang-Emit-build-time-warning-for-Clang-pri_patch.txt)

**Comment by nathanchance** [comment]
I still do not really like the idea of doing this yet but if there is a general consensus then I will not argue.

What's the point of the first patch? We can use `CONFIG_CC_IS_CLANG` and `CONFIG_CLANG_VERSION` in C, Kconfig, and make files, why do we need `cc-is-clang`, `clang-version`, etc?

The second patch's title needs to be improved I think. We aren't warning about an older clang, we're straight up breaking the build.

I've started working on the testing framework that I mentioned in the post, you can see it [here](https://github.com/nathanchance/llvm-kernel-testing). Once LLVM 10.0.1-rc1 gets posted, I'll be able to see how it truly works as I have definitely tripped over quite a few bugs in LLVM 10.0.0. Until that happens though, I do not feel comfortable limiting ourselves with a patch like this. We have to ensure that LLVM releases do not regress before we can start saying "hey, this version will DEFINITELY work".

**Comment by dileks** [comment]
New discussion in thread "[PATCH] Restore gcc check in mips asm/unroll.h"

[1] https://lore.kernel.org/r/20200709221102.323356-1-cesarb@cesarb.eti.br/

**Comment by dileks** [comment]
From @nickdesaulniers 

"Documentation: add minimum clang/llvm version"

[1] https://lore.kernel.org/r/20200825222552.3113760-1-ndesaulniers@google.com/

**Comment by nickdesaulniers** [comment]
https://lore.kernel.org/lkml/20200826201420.3414123-1-ndesaulniers@google.com/T/#u

**Comment by nickdesaulniers** [comment]
Jon Corbet picked up the Documentation patch: https://lore.kernel.org/lkml/20200831163048.0cbeca5d@lwn.net/

I need to resend the header.

**Comment by nickdesaulniers** [comment]
v2: https://lore.kernel.org/lkml/20200901002326.1137289-1-ndesaulniers@google.com/T/#t

**Comment by nickdesaulniers** [comment]
v3: https://lore.kernel.org/lkml/20200902225911.209899-1-ndesaulniers@google.com/T/#t

**Comment by nathanchance** [comment]
We have committed to LLVM 10.0.1 for the time being and that series of patches has landed in 5.10 now. Closing this for now.

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/?h=c8db3b0a7ba7614f761f309d6aa7499127b18a0b
https://git.kernel.org/linus/3519c4d6e08ea695658b961829d4a603acf8e28a
